translator/opencensus/oc_to_internal.go

```diff
@@ -0,0 +1,405 @@
+	if ocEvents == nil {
+		return
+	}
+
+	droppedCount = uint32(ocEvents.DroppedMessageEventsCount + ocEvents.DroppedAnnotationsCount)
+
+	evenCount := len(ocEvents.TimeEvent)
+	if evenCount == 0 {
+		return
+	}
+
+	// Create slice that holds pointers and another slice that holds structs at once
+	// to avoid individual allocations of structs.
+	ptrs = make([]*data.SpanEvent, 0, evenCount)
+	content := data.NewSpanEventSlice(evenCount)
+
+	for i, ocEvent := range ocEvents.TimeEvent {
+		if ocEvent == nil {
+			continue
+		}
+
+		// Point one element in slice of pointers to an element in slice of structs.
+		event := &content[i]
+		ptrs = append(ptrs, event)
+
+		event.SetTimestamp(timestampToUnixnano(ocEvent.Time))
+
+		switch teValue := ocEvent.Value.(type) {
+		case *octrace.Span_TimeEvent_Annotation_:
+			if teValue.Annotation != nil {
+				event.SetName(truncableStringToStr(teValue.Annotation.Description))
+				attrs := ocAttrsToInternal(teValue.Annotation.Attributes)
+				event.SetAttributes(attrs)
+			}
+
+		case *octrace.Span_TimeEvent_MessageEvent_:
+			event.SetAttributes(ocMessageEventToInternalAttrs(teValue.MessageEvent))
+
+		default:
+			event.SetName("An unknown OpenCensus TimeEvent type was detected when translating")
+		}
+	}
+	return
+}
+
+func ocLinksToInternal(ocLinks *octrace.Span_Links) (ptrs []*data.SpanLink, droppedCount uint32) {
+	if ocLinks == nil {
+		return
+	}
+
+	droppedCount = uint32(ocLinks.DroppedLinksCount)
+
+	linkCount := len(ocLinks.Link)
+	if linkCount == 0 {
+		return
+	}
+
+	// Create slice that holds pointers and another slice that holds structs at once
+	// to avoid individual allocations of structs.
+	ptrs = make([]*data.SpanLink, 0, linkCount)
+	content := data.NewSpanLinkSlice(linkCount)
+
+	for i, ocLink := range ocLinks.Link {
+		if ocLink == nil {
+			continue
+		}
+
+		// Point one element in slice of pointers to an element in slice of structs.
+		link := &content[i]
+		ptrs = append(ptrs, link)
+
+		attrs := ocAttrsToInternal(ocLink.Attributes)
+		link.SetTraceID(data.TraceIDFromBytes(ocLink.TraceId))
+		link.SetSpanID(data.SpanIDFromBytes(ocLink.SpanId))
+		link.SetTraceState(ocTraceStateToInternal(ocLink.Tracestate))
+		link.SetAttributes(attrs)
+	}
+	return ptrs, droppedCount
+}
+
+func ocMessageEventToInternalAttrs(msgEvent *octrace.Span_TimeEvent_MessageEvent) data.Attributes {
+	if msgEvent == nil {
+		return data.Attributes{}
+	}
+
+	attrs := data.NewAttributeValueSlice(4)
+	attrs[0].SetString(msgEvent.Type.String())
+	attrs[1].SetInt(int64(msgEvent.Id))
+	attrs[2].SetInt(int64(msgEvent.UncompressedSize))
+	attrs[3].SetInt(int64(msgEvent.CompressedSize))
+
+	return data.NewAttributes(
+		map[string]data.AttributeValue{
+			ocTimeEventMessageEventType:  attrs[0],
+			ocTimeEventMessageEventID:    attrs[1],
+			ocTimeEventMessageEventUSize: attrs[2],
+			ocTimeEventMessageEventCSize: attrs[3],
+		},
+		0)
+}
+
+func truncableStringToStr(ts *octrace.TruncatableString) string {
```

Comment:
nit: Duplicate of https://github.com/open-telemetry/opentelemetry-collector/blob/c0f4a5a70d825c2466f98034bcfed9293a61c870/translator/trace/oc_to_otlp.go#L316-L321
but it's trivial, so might be preferred to keep them separate.

Reply:
I think it is tiny, let it be.